concourse/steps/release.py

```
@@ -23,19 +28,303 @@
+                self.git_helper._pop_stash()
+
+    def _invoke_release_callback(
+        self,
+        release_commit_callback,
+        repo_dir,
+        release_version,
+    ):
+        callback_env = os.environ.copy()
+        callback_env['REPO_DIR'] = repo_dir
+        callback_env['EFFECTIVE_VERSION'] = release_version
+
+        subprocess.run(
+            [release_commit_callback],
+            check=True,
+            env=callback_env,
+        )
+
+    def revert(self):
+        # TODO: revert commit
+        pass
+
+
+class ReleaseTagStep(TransactionalStep):
+    def __init__(
+        self,
+        github_helper:GitHubRepositoryHelper,
+        release_version: str,
+        author_name: str,
+        author_email: str,
+        release_context: ReleaseContext,
+    ):
+        super().__init__(release_context)
+        self.github_helper = not_none(github_helper)
+        self.release_version = not_empty(release_version)
+        self.author_name = not_empty(author_name)
+        self.author_email = not_empty(author_email)
+
+    def validate(self):
+        info(f'validating { self.name()}')
+        # if a tag with the given release version already exists github will not let us
+        # create another one
+        if self.github_helper.tag_exists(tag_name=self.release_version):
+            raise RuntimeError(
+                f"Cannot create tag '{self.release_version}' in preparation for release: "
+                "Tag already exists"
+            )
+
+    def name(self):
+        return "Release Tag Creator"
+
+    def apply(self):
+        info('Creating tag for release')
+        self.github_helper.create_tag(
+            tag_name=self.release_version,
+            tag_message=f'Release {self.release_version}',
+            repository_reference=self.release_context.release_commit_sha(),
+            author_name=self.author_name,
+            author_email=self.author_email
+        )
+
+    def revert(self):
+        # TODO: Delete tag
```

ditto - do not silently return